File: spC++/src/Implement/Services/TableOfContentsUpdate.cpp
#include "Headers/Services/TableOfContentsUpdate.hpp"

void TableOfContentsUpdate::visitBook(Book *book) {
    for (Visitee* element: book->getChildren()) {
        element->accept(this);
    }
}

void TableOfContentsUpdate::visitSection(Section *section) {
    tableOfContents->addEntry(section->getTitle());
    for (IElement* element: section->getChildren()) {
        element->accept(this);
    }
}

void TableOfContentsUpdate::visitToC(TableOfContents *tableOfContents) {
    std::cout << "";
}

void TableOfContentsUpdate::visitParagraph(Paragraph *paragraph) {
    tableOfContents->addEntry(nullptr);
}

void TableOfContentsUpdate::visitIProxy(ImageProxy *imageProxy) {
    imageProxy->loadImage().accept(this);
}

void TableOfContentsUpdate::visitImage(Image *image) {
    tableOfContents->addEntry(nullptr);
}

void TableOfContentsUpdate::visitTable(Table *table) {
    tableOfContents->addEntry(nullptr);
}

void TableOfContentsUpdate::visitAuthor(Author *author) {
    return;
}

File: Java-Labs/lab/src/main/java/sp/lab/Services/TableOfContentsUpdate.java
package sp.lab.Services;

import sp.lab.Models.Book;
import sp.lab.Models.IVisitor;
import sp.lab.Models.Image;
import sp.lab.Models.ImageProxy;
import sp.lab.Models.Paragraph;
import sp.lab.Models.Section;
import sp.lab.Models.Table;
import sp.lab.Models.TableOfContents;

public class TableOfContentsUpdate implements IVisitor{
    private TableOfContents tableOfContents;
    private int pageIndex;

    public TableOfContentsUpdate() {
        this.tableOfContents = new TableOfContents();
        this.pageIndex = 0;
    }

    @Override
    public void visitBook(Book book) {
        this.pageIndex++;
    }

    @Override
    public void visitSection(Section section) {
        String entry = section.getTitle() + ' ';
        while(entry.length() <= 20) {
            entry += '.';
        }
        entry += ' ' + this.pageIndex;
        
        this.tableOfContents.addEntry(entry);
    }

    @Override
    public void visitParagraph(Paragraph paragraph) {
        this.pageIndex++;
    }

    @Override
    public void visitImageProxy(ImageProxy imageProxy) {
        this.pageIndex++;
    }

    @Override
    public void visitImage(Image image) {
        this.pageIndex++;
    }

    @Override
    public void visitTable(Table table) {
        this.pageIndex++;
    }

    @Override
    public void visitTableOfContents(TableOfContents tableOfContents) {}

    public TableOfContents getTableOfContents() {
        return this.tableOfContents;
    }
}
